kernel/Utilities/BasisFunctions2DNedelec.cpp

```diff
@@ -37,77 +37,229 @@
  */
 
 #include "BasisFunctions2DNedelec.h"
-#include "helperFunctions.h"
+
+#include <utility>
+
+#include "Base/BaseBasisFunction.h"
 #include "Base/BasisFunctionSet.h"
-#include "Geometry/ReferenceTriangle.h"
 #include "Geometry/PointReference.h"
+#include "Geometry/ReferenceTriangle.h"
+
+#include "BasisFunctions1DH1ConformingLine.h"
+#include "BasisFunctions2DH1ConformingTriangle.h"
 
 namespace hpgem {
 
 namespace Utilities {
-namespace {
-LinearAlgebra::SmallVector<2> baricentricDeriv(std::size_t node) {
-    LinearAlgebra::SmallVector<2> ret;
-    if (node == 0) {
-        ret[0] = -1;
-        ret[1] = -1;
+
+/// Edge based basis functions of the form form q_p(t) F(x,y).
+///
+/// The function F(x,y) gives the direction of the vector field and consists of
+/// two linear functions. For each side these are constructed so that it has a
+/// constant tangential trace along that side and zero tangential trace along
+/// the two other sides.
+///
+/// The function q_p(t) is shared between the sides and 'modulates' the vector
+/// field by a polynomial of order p-1. The coordinate t is linear on the side
+/// to which the basis function is associated.
+class BasisCurlEdgeNedelec2D : public Base::BaseBasisFunction {
+   public:
+    BasisCurlEdgeNedelec2D(
+        std::size_t side,
+        std::shared_ptr<const Base::BaseBasisFunction> modulator)
+        : side_(side), modulator_(std::move(modulator)) {
+
+        logger.assert_debug(side <= 2, "Triangle has only three sides");
+    }
+
+    void eval(const Geometry::PointReference<2>& p,
+              LinearAlgebra::SmallVector<2>& ret) const override {
+        evalSideFunction(p, ret);
+        Geometry::PointReference<1> coord;
+        coord[0] = evalModulatorCoord(p);
+        ret *= modulator_->eval(coord);
+    }
+
+    LinearAlgebra::SmallVector<2> evalCurl(
+        const Geometry::PointReference<2>& p) const override;
+
+   private:
+    /// The side function F(x,y)
+    void evalSideFunction(const Geometry::PointReference<2>& p,
```

Comment:
probably because od the interface but it would be much nicer to return ret instead of passing it in as a non-const ref

Reply:
I would keep it consistent with all the interface for `eval`, where it is also passed as out-parameter.

Reply:
okay